libraries/platforms/dromajo-vcs/bsg_manycore_simulator.hpp

```diff
@@ -0,0 +1,111 @@
+// of conditions and the following disclaimer.
+// 
+// Redistributions in binary form must reproduce the above copyright notice, this
+// list of conditions and the following disclaimer in the documentation and/or
+// other materials provided with the distribution.
+// 
+// Neither the name of the copyright holder nor the names of its contributors may
+// be used to endorse or promote products derived from this software without
+// specific prior written permission.
+// 
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+// ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+// WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR
+// ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+// (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+// LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+// ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+// This file implements the SimulationWrapper object.
+
+#ifndef __BSG_MANYCORE_SIMULATOR_HPP
+#define __BSG_MANYCORE_SIMULATOR_HPP
+
+#include <stdlib.h>
+#include <stdio.h>
+#include <stdarg.h>
+#include <stdint.h>
+#include <unistd.h>
+#include <string.h>
+
+#include <string>
+
+#include <dromajo_cosim.h>
+#include <dromajo_manycore.h>
+
+#include <bsg_nonsynth_dpi_manycore.hpp>
+#include <bsg_nonsynth_dpi_errno.hpp>
+#include <bsg_nonsynth_dpi_cycle_counter.hpp>
+#include <bsg_nonsynth_dpi_clock_gen.hpp>
+
+#include <bsg_manycore.h>
+#include <bsg_manycore_config.h>
+#include <bsg_manycore_printing.h>
+
+#include <bsg_manycore_regression.h>
+
+#include <bp_hb_platform.h>
+
+// Define invisible library functions
+declare_hb_mc_get_bits
+declare_bsg_printing
+
+#define NUM_DROMAJO_INSTR_PER_TICK 1000
+
+class SimulationWrapper{
+  // This is the generic pointer for implementation-specific
+  // simulator details. In Verilator, this is
+  // Vmanycore_tb_top. In VCS this is the scope
+  // for DPI.
+  void *top = nullptr;
+  std::string *root;
+  // Pointer to the dromajo instance
+  dromajo_cosim_state_t *dromajo;
+  // Pointer to the manycore DPI controller
+  bsg_nonsynth_dpi::dpi_manycore<HB_MC_CONFIG_MAX> *dpi;
+
+public:
+  SimulationWrapper();
+  ~SimulationWrapper();
+
+  // Change the assertion state. 
+
+  // When Verilator simulation starts, we want to disable
+  // assertion because it is a two-state simulator and the lack
+  // of z/x may cause erroneous assertions.
+  //
+  // This does not need to be implemented in 4-state simulators
+  // like VCS
+  void assertOn(bool val);
+  
+  std::string getRoot();
+
+  // Causes time to proceed by 1 unit
+  void advance_time();
```

Comment:
By one unit, or to the next clock edge of the simulation interface?

This is why bsg_dpi_next is named as it is. It goes to the _next_ event.

Reply:
Oh, that should be to the next clock edge. This function just calls `bsg_dpi_next` by setting the correct svScope. I'll update the comment.

Reply:
FWIW, There are other places with the same comment

Reply:
I searched and modified the comment wherever I saw it. If there are any lingering instances you find, please let me know.